File: dotnet/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example2/DateRangeIterator.cs
namespace IteratorDesignPattern.Example2;

public class DateRangeIterator : IIterator<Order>
{
    private readonly OrderCollection _collection;
    private readonly DateTime _startDate;
    private readonly DateTime _endDate;
    private int _currentIndex = 0;

    public DateRangeIterator(OrderCollection collection, DateTime startDate, DateTime endDate)
    {
        _collection = collection;
        _startDate = startDate;
        _endDate = endDate;
    }

    public bool HasNext()
    {
        while (_currentIndex < _collection.Count)
        {
            var orderDate = _collection[_currentIndex].OrderDate.Date;
            if (orderDate >= _startDate.Date && orderDate <= _endDate.Date)
            {
                return true;
            }
            _currentIndex++;
        }
        return false;
    }

    public Order Next()
    {
        if (!HasNext())
        {
            throw new InvalidOperationException("No more orders in the date range.");
        }

        return _collection[_currentIndex++];
    }

    public void Reset()
    {
        _currentIndex = 0;
    }
}

File: java/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example2/src/main/java/behavioral/iterator/example2/DateRangeIterator.java
package behavioral.iterator.example2;

import java.time.LocalDate;

public class DateRangeIterator implements IIterator<Order> {
    private final OrderCollection collection;
    private final LocalDate startDate;
    private final LocalDate endDate;
    private int currentIndex = 0;

    public DateRangeIterator(OrderCollection collection, LocalDate startDate, LocalDate endDate) {
        this.collection = collection;
        this.startDate = startDate;
        this.endDate = endDate;
    }

    @Override
    public boolean hasNext() {
        while (currentIndex < collection.getCount()) {
            LocalDate date = collection.get(currentIndex).getOrderDate();
            if (!date.isBefore(startDate) && !date.isAfter(endDate)) return true;
            currentIndex++;
        }
        return false;
    }

    @Override
    public Order next() {
        if (!hasNext()) throw new java.util.NoSuchElementException();
        return collection.get(currentIndex++);
    }

    @Override
    public void reset() { currentIndex = 0; }
}
